drivers/ec_host_cmd_periph/ec_host_cmd_periph_espi.c

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (c) 2022 Google LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT zephyr_ec_host_cmd_periph_espi
+
+#include <string.h>
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/ec_host_cmd_periph/ec_host_cmd_periph.h>
+#include <zephyr/drivers/espi.h>
+#include <zephyr/mgmt/ec_host_cmd.h>
+
+BUILD_ASSERT(DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) == 1, "Invalid number of eSPI peripherals");
+
+#define ESPI_BUS    DT_PHANDLE(DT_DRV_INST(0), bus)
+#define ESPI_DEVICE DEVICE_DT_GET(ESPI_BUS)
+
+struct ec_host_cmd_periph_espi_data {
+	struct k_sem handler_owns;
+	struct k_sem dev_owns;
+	uint32_t rx_buffer_len;
+	struct espi_callback espi_cb;
+	uint8_t *espi_shm;
+};
+
+static void ec_host_cmd_periph_espi_handler(const struct device *dev, struct espi_callback *cb,
+					    struct espi_event espi_evt)
+{
+	struct ec_host_cmd_periph_espi_data *data =
+		CONTAINER_OF(cb, struct ec_host_cmd_periph_espi_data, espi_cb);
+	uint16_t event_type = (uint16_t)espi_evt.evt_details;
+
+	if (event_type != ESPI_PERIPHERAL_EC_HOST_CMD) {
+		return;
+	}
+
+	if (k_sem_take(&data->dev_owns, K_NO_WAIT) != 0) {
+		int res = EC_HOST_CMD_IN_PROGRESS;
+
+		espi_write_lpc_request(ESPI_DEVICE, ECUSTOM_HOST_CMD_SEND_RESULT, &res);
+		return;
+	}
+
+	k_sem_give(&data->handler_owns);
+}
+
+int ec_host_cmd_periph_espi_init(const struct device *dev, struct ec_host_cmd_periph_rx_ctx *rx_ctx)
+{
+	struct ec_host_cmd_periph_espi_data *data = dev->data;
+
+	if (rx_ctx == NULL) {
+		return -EINVAL;
+	}
+
+	rx_ctx->buf = data->espi_shm;
+	rx_ctx->len = &data->rx_buffer_len;
+	rx_ctx->dev_owns = &data->dev_owns;
+	rx_ctx->handler_owns = &data->handler_owns;
+
+	return 0;
+}
+
+int ec_host_cmd_periph_espi_send(const struct device *dev,
+				 const struct ec_host_cmd_periph_tx_buf *buf)
+{
+	struct ec_host_cmd_periph_espi_data *data = dev->data;
+	struct ec_host_cmd_response_header *resp_hdr =
+		(struct ec_host_cmd_response_header *)buf->buf;
```

Comment:
I think the casting is not needed here since we are starting from a void *

Reply:
Right, thanks!